src/query/ts/m3db/encoded_step_iterator.go

```diff
@@ -0,0 +1,236 @@
+}
+
+type encodedStepIter struct {
+	mu           sync.RWMutex
+	lastBlock    bool
+	started      bool
+	currentTime  time.Time
+	err          error
+	bounds       models.Bounds
+	meta         block.Metadata
+	seriesMeta   []block.SeriesMeta
+	seriesIters  []encoding.SeriesIterator
+	seriesPeek   []peekValue
+	consolidator *consolidators.StepLookbackConsolidator
+}
+
+func (b *encodedBlock) stepIter() block.StepIter {
+	cs := b.consolidation
+	consolidator := consolidators.NewStepLookbackConsolidator(
+		time.Minute,
+		cs.bounds.StepSize,
+		cs.currentTime,
+		len(b.seriesBlockIterators),
+		cs.consolidationFn,
+	)
+	return &encodedStepIter{
+		lastBlock:    b.lastBlock,
+		currentTime:  cs.currentTime,
+		bounds:       cs.bounds,
+		meta:         b.meta,
+		seriesMeta:   b.seriesMetas,
+		seriesIters:  b.seriesBlockIterators,
+		consolidator: consolidator,
+	}
+}
+
+type encodedStep struct {
+	time   time.Time
+	values []float64
+}
+
+func (s *encodedStep) Time() time.Time   { return s.time }
+func (s *encodedStep) Values() []float64 { return s.values }
+
+func (it *encodedStepIter) Current() (block.Step, error) {
+	it.mu.RLock()
+	currentTime := it.currentTime
+	if !it.started || currentTime.After(it.bounds.End()) {
+		it.mu.RUnlock()
+		return nil, fmt.Errorf("out of bounds")
+	}
+
+	if it.err != nil {
+		it.mu.RUnlock()
+		return nil, it.err
+	}
+
+	step := &encodedStep{
+		time:   currentTime,
+		values: it.consolidator.ConsolidateAndMoveToNext(),
+	}
+
+	it.mu.RUnlock()
+	return step, nil
+}
+
+// Moves to the next consolidated step for the i-th series in the block,
+// populating the consolidator for that step. Will keep reading values
+// until either hitting the next step boundary and returning, or until
+// encountering a value beyond the boundary, at which point it adds it
+// to a stored peeked value that is consumed on the next pass.
+func (it *encodedStepIter) nextConsolidatedForStep(i int) {
+	peek := it.seriesPeek[i]
+	if peek.finished {
+		// No next value in this iterator
+		return
+	}
+
+	if peek.started {
+		point := peek.point
+		if point.Timestamp.After(it.currentTime) {
+			// This point exists further than the current step
+			// There are next values, but this point should be NaN
+			return
+		}
+
+		// Currently at a potentially viable data point.
+		// Record previously peeked value, and all potentially valid
+		// values, then apply consolidation function to them to get the
+		// consolidated point.
+		it.consolidator.AddPointForIterator(point, i)
+		// clear peeked point.
+		it.seriesPeek[i].started = false
+		// If at boundary, add the point as the current value.
+		if point.Timestamp.Equal(it.currentTime) {
+			return
+		}
+	}
+
+	iter := it.seriesIters[i]
+	// Read through iterator until finding a data point outside of the
+	// range of this consolidated step; then consolidate those points into
+	// a value, set the next peek value, and return true.
+	for iter.Next() {
+		dp, _, _ := iter.Current()
```

Comment:
Check error?

Reply:
If seriesIter fails on a `Next` it will return false, so it will not enter this codeblock. Instead, error is checked after the `Next` block is finished